crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/AuthenticationMethodRetriever.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.registry;
+
+import com.google.api.client.http.HttpMethods;
+import com.google.api.client.http.HttpResponseException;
+import com.google.api.client.http.HttpStatusCodes;
+import com.google.cloud.tools.crepecake.http.BlobHttpContent;
+import com.google.cloud.tools.crepecake.http.Response;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/** Retrieves the {@code WWW-Authenticate} header from the registry API. */
+class AuthenticationMethodRetriever implements RegistryEndpointProvider<RegistryAuthenticator> {
+
+  private final RegistryEndpointProperties registryEndpointProperties;
+
+  @Nullable
+  @Override
+  public BlobHttpContent getContent() {
+    return null;
+  }
+
+  @Override
+  public List<String> getAccept() {
+    return Collections.emptyList();
+  }
+
+  @Override
+  public RegistryAuthenticator handleResponse(Response response) throws RegistryErrorException {
+    throw new RegistryErrorExceptionBuilder(getActionDescription())
```

Comment:
Maybe I'm missing something, but this seems to be always throwing this exception?

Reply:
Yea, the RegistryAuthenticator expects a `401 Unauthorized` HTTP error, so if it receives a 200 status response (handled by `handleResponse`), it throws an exception.

Reply:
Oh I see, this just picks up the authentication method... is this really the only way to do it? Isn't there an endpoint to hit to ask?

Reply:
Yeah, this is doing the authentication according to https://docs.docker.com/registry/spec/auth/token/, where:

1) Get the authentication method from the WWW-Authenticate header
2) Authenticate with the provided method (done by `RegistryAuthenticator`)

Reply:
Damn that's so silly. okay :smile:

Reply:
Here, we are doing it once before all pull calls. But, we might want to have this be done every time we receive a 401 Unauthorized, so I filed https://github.com/GoogleCloudPlatform/minikube-build-tools-for-java/issues/96

Reply:
Not sure if you linked to that somewhere, but it'd be nice to include in the code so someone looking at this in the future will understand.

Reply:
just kdiding I found it in the code...

Reply:
Yeah, it is linked in `RegistryAuthenticator`. `AuthenticationMethodRetriever` just does standard `WWW-Authenticate` header retrieval.